### PR TITLE
lkl: Replace threads_counter_lock with atomic ops

### PR DESCRIPTION
**Pull request by pscollins**
Although the __sync_fetch_and_\* ops are a GCC extension, we already rely
on them in arch/lkl/kernel/syscalls.c, so there's no loss of portability
here.

Signed-off-by: Patrick Collins pscollins@google.com

[Review on Reviewable](https://reviewable.io/reviews/lkl/linux/114)

### CODE REVIEW

File: arch/lkl/kernel/threads.c [hunk 1]
```diff
@@ -123,7 +97,7 @@ struct task_struct *__switch_to(struct task_struct *prev,
 
 	if (ei.dead) {
 		lkl_ops->sem_free(ei.sched_sem);
-		threads_counter_dec();
+		__sync_fetch_and_add(&threads_counter, -1);
```
Comment (stfairy) [inline arch/lkl/kernel/threads.c]: __sync_fetch_and_sub

Comment (pscollins) [inline arch/lkl/kernel/threads.c]: @stfairy Whoops, I missed that in the docs, thanks! Done.

**Comment by thehajime** [comment]
while it may be true that we don't loose any portability issue, this dependency to the GCC extension looks like not mandatory here.

do you have any particular issue if we don't have this patch ? if there is, it's nice to have the patch but would like to see what it is.

we will face an issue if some underlying environment doesn't have this extension. 
for me it's the same discussion with minimum size of ``host_ops``` I mentioned in #85.
(you know I'm so much conservative here)

**Comment by thehajime** [comment]
sigh. 

I don't really understand why a PR is closed without resolving an objection.

**Comment by opurdila** [comment]
@thehajime sorry, I have missed your comment because I only checked the commit comments and not the PR comments.

As @pscollins mentioned, we are already using these extensions for irq and syscall handling and I don't see a way to avoid them there.

**Comment by thehajime** [comment]
okay @opurdila.

> As @pscollins mentioned, we are already using these extensions for irq and syscall handling and I don't see a way to avoid them there.

I've understood irq and syscall parts but didn't see whether this particular commit _really_ needs the GCC extension.

**Comment by pscollins** [comment]
@thehajime What's the benefit to having multiple implementations of the same thing scattered across the system? We already have a dependency on this extension so there's no gain in portability.

**Comment by thehajime** [comment]
> @thehajime What's the benefit to having multiple implementations of the same thing scattered across the system? We already have a dependency on this extension so there's no gain in portability.

@pscollins I agree with you having multiple impl. of the same procedures is problematic.
I'm against not only this PR but also atomic built-ins for irq and syscall.

if there is reasonable motivation for this patch like there is _visible_ performance boost for instance, I'm fine.

why I'm complaining here (and atomic built-ins) is simply for broader host support in a future:  I have a new host ops in my hand with a compiler which doesn't have atomic_ ops and sync_ xxx even though the gcc version is 5.2.0. The environment doesn't have OS and libc either. 
(I'm talking about ```arm-none-eabi-gcc```, gcc arm embedded)

such a restricted environment _can_ be a host of LKL, with a small footprint of software stack.

right now I need to use libatomic (available at https://gcc.gnu.org/wiki/Atomic/GCCMM) but it's not shipped with older gcc (the web page says it's after 4.8 but my arm-none-eabi-gcc with 5.2.0 on Fedora 23 doesn't have...). so far it's complicated and looks not always useful. other hosts may have similar issue.

If we can do something with a different way without atomic ops, I feel it's a bad signal in this sense (i.e., more host ops support). If we can't avoid (I assume the irq and syscall case are in this situation), then we need to use the atomic ops and wrap with ifdefs for unsupported platform. 


We are library. We aren't at the bottom of software stack nor top: I think being a generic implementation as much as possible will bring better support for various environments.

I can see those portabilities are fully incompatible with any level of optimizations (e.g., performance optimization) so, this is a design choice where we would go for.

would like to see more opinions.

**Comment by opurdila** [comment]
> I'm against not only this PR but also atomic built-ins for irq and syscall.

The main reason for switching to the current irq handling mechanism was to fix a deadlock which didn't allow us to trigger interrupts from softirq or irq context.

> why I'm complaining here (and atomic built-ins) is simply for broader host support in a future: I have a new host ops in my hand with a compiler which doesn't have atomic_ ops and sync_ xxx even though the gcc version is 5.2.0. The environment doesn't have OS and libc either. 
(I'm talking about arm-none-eabi-gcc, gcc arm embedded)

Does it support threads and semaphores?

What we can do is to move the atomic ops to host_ops and provide (common) implementations for atomics using mutexes. Thus, host that don't have/provide atomics in host_ops will use the default mutex based implementation.

**Comment by thehajime** [comment]
> I'm against not only this PR but also atomic built-ins for irq and syscall.
> 
> The main reason for switching to the current irq handling mechanism was to fix a deadlock which didn't allow us to trigger interrupts from softirq or irq context.
> 

I see the point.
Do you have a test code (or code snippet) that can reproduce the deadlock ? 

I'd like to understand more this issue. 

> why I'm complaining here (and atomic built-ins) is simply for broader host support in a future: I have a new host ops in my hand with a compiler which doesn't have atomic_ ops and sync_ xxx even though the gcc version is 5.2.0. The environment doesn't have OS and libc either. 
> (I'm talking about arm-none-eabi-gcc, gcc arm embedded)
> 
> Does it support threads and semaphores?
> 
There is ucontext based thread. 

> What we can do is to move the atomic ops to host_ops and provide (common) implementations for atomics using mutexes. Thus, host that don't have/provide atomics in host_ops will use the default mutex based implementation.
> 
It would make sense but I'm afraid atomic ops using mutex again has the deadlock issue....

**Comment by pscollins** [comment]
> I'm against not only this PR but also atomic built-ins for irq and syscall.

Atomic ops are (in my opinion) the most fundamental synchronization primitive --- you can build locks and semaphores with proper semantics out of atomic ops but not vice-versa. A system without atomic ops has serious, serious issues. 

> There is ucontext based thread.

If there are no mutexes, the current implementation won't work anyway? And the LKL scheduler relies on the host scheduler to behave like a normal OS when it raises and lowers semaphores, so I think `__switch_to` would probably be broken on this system.


But it's a valid concern, anyway; maybe this is justification for the refactor I described in the `calloc` PR.

**Pull request merged**